src/System.Diagnostics.TraceSource/src/System/Diagnostics/SwitchAttribute.cs

```diff
@@ -0,0 +1,92 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Reflection;
+using System.Collections;
+
+namespace System.Diagnostics
+{
+    [AttributeUsage(AttributeTargets.Assembly | AttributeTargets.Class | AttributeTargets.Constructor |
+                    AttributeTargets.Event | AttributeTargets.Method | AttributeTargets.Property)]
+    public sealed class SwitchAttribute : Attribute
+    {
+        private Type _type;
+        private string _name;
+        
+        public SwitchAttribute(string switchName, Type switchType)
+        {
+            SwitchName = switchName;
+            SwitchType = switchType;
+        }
+
+        public string SwitchName
+        {
+            get { return _name; }
+            set
+            {
+                if (value == null)
+                    throw new ArgumentNullException("value");
+                if (value.Length == 0)
+                    throw new ArgumentException(SR.Format(SR.InvalidNullEmptyArgument, "value"), "value");
+
+                _name = value;
+            }
+        }
+
+        public Type SwitchType
+        {
+            get { return _type; }
+            set
+            {
+                if (value == null)
+                    throw new ArgumentNullException("value");
+                _type = value;
+            }
+        }
+
+        public string SwitchDescription { get; set; }
+
+        public static SwitchAttribute[] GetAll(Assembly assembly)
+        {
+            if (assembly == null)
+                throw new ArgumentNullException("assembly");
+
+            ArrayList switchAttribs = new ArrayList();
```

Comment:
Because the reflection API returns object[] a bit more convoluted [code is needed](https://github.com/dotnet/corefx/pull/12527/commits/dbbfcb673780c14c4249e3f9614dd9c96ccd6d2d).

The other option is to use a List<object> instead of an ArrayList (which removes the need for the CastTo method).

I addressed the rest of the feedback.